### PR TITLE
Update windows azure plugin

### PR DESCRIPTION
**Pull request by roguishmountain**
Major changes are:
- Adding the capability to use managed artifacts the same way the built in manage artifacts plugin use it
- Updating the Azure SDK to the version that is currently the most recent
- Fixing the links on the project page to download properly. The code was not getting the correct shared access signature
- Speed up downloads. Due to having managed artifacts the plugin no longer needs to search the entire container for the correct blobs
- Updates to the code to be in line with changes to the Jenkins API
- Improved the output to the Jenkins REST API

### CODE REVIEW

**Comment by martinsawicki** [comment]
@snallami, any chance you could take a look?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStorageClient.java [hunk 1]
```diff
@@ -303,12 +309,31 @@ private static void upload(TaskListener listener, Launcher launcher, CloudBlockB
...
+     * @param expFP File Path in ant glob syntax relative to CI tool workspace.
+     * @param expVP Virtual Path of blob container.
+     * @param excludeFP File Path in ant glob syntax to exclude from upload
+     * @param uploadType upload file type
+     * @param individualBlobs blobs from build
+     * @param archiveBlobs blobs from build in archive files
+     * @param cleanUpContainer if container is cleaned
+     * @return filesUploaded number of files that are uploaded.
+     * @throws WAStorageException throws exception
+>>>>>>> 22d2ec7... remove unneeded manageArtifacts flag and fix excessive if statements
```
Comment (duncanmak) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStorageClient.java]: There's a merge marker that you should remove

**Comment by arroyc** [comment]
looking into it ...

**Comment by clguimanMSFT** [comment]
1. How can the user use the old behavior and download an artifact from a container (using the name), not just from another Jenkins Job's artifact?
2. Is there some upgrade path from a previous version?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 2]
```diff
@@ -358,6 +341,54 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath ws, @Nonnull Launc
 		}
 	}
 
+	private void validateData(Run<?, ?> run,
+			TaskListener listener, StorageAccountInfo storageAccount,
+			String expContainerName) throws IOException, InterruptedException {
```
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: What's the use of container name ?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 3]
```diff
@@ -200,32 +201,15 @@ public void perform(Run<?,?> run, FilePath filePath, Launcher launcher, TaskList
...
-						run.setResult(Result.UNSTABLE);
-					} else {
-						listener.getLogger()
-								.println(
-										Messages.AzureStorageBuilder_files_downloaded_count(filesDownloaded));
-					}
-				} else if (source instanceof FreeStyleBuild) {
-					downloadArtifacts(source, run, launcher, expIncludePattern, expExcludePattern, listener, strAcc);
-				} else if (source instanceof MatrixBuild) {
+				if (source instanceof MatrixBuild) {
```
Comment (arroyc) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: we should also keep this old way of downloading bits from container

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java (hunk 2 above)
Comment (roguishmountain) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: @arroyc This diff is outdated. Validation of the container name is added in a later commit.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 3 above)
Comment (roguishmountain) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: @arroyc What do you mean by "this old way"?

**Comment by roguishmountain** [comment]
@clguimanMSFT 
1. The new behavior is in line with other artifact managers such as the Jenkins internal artifact manager. It uses the build selectors from https://github.com/jenkinsci/copyartifact-plugin. An alternative could be to upload an artifact from a job and then use the "download from specific build" option.
2. On the UI side there are only two changes: a) to add the project name and b) the build selector. There is also a way to mark a plugin incompatible with prior versions if needed.

**Comment by clguimanMSFT** [comment]
I think there are valid use cases when the users just wants to download a file from a container without having a previous job creating that artifact.
We should definitely support both scenarios (maybe add a drop-down to select the type of download : from container or from another job). In this case we must make sure that the latest plugin will be able to migrate existing configurations

**Comment by clguimanMSFT** [comment]
Rebased the code and added extra stuff in [PR-32](https://github.com/jenkinsci/windows-azure-storage-plugin/pull/32)
If everyone's ok I'll close this PR and we can move our conversation to PR-32

**Comment by clguimanMSFT** [comment]
@roguishmountain can i close this PR and move to PR-32?

**Comment by roguishmountain** [comment]
@clguimanMSFT I'm about to update the help files as I mentioned on PR-32. After the commit all the changes will be on this PR and it would probably be easier to merge this one.

**Comment by clguimanMSFT** [comment]
It's not easy to merge this one, that's why I created a new one so I could merge all the conflicts. If you want to handle the merge be my guest :)

**Comment by roguishmountain** [comment]
@clguimanMSFT You can use your branch, just add the commit I just made. Also, https://github.com/Microsoft/azure-docs/blob/master/articles/storage/storage-java-jenkins-continuous-integration-solution.md will probably need to be updated to reflect the changes in the plugin.

**Comment by clguimanMSFT** [comment]
Done. I'll close the PR. We'll update the help file once the feature makes it in a public release. We have more features coming in soon, so we don't release just with this one

**Pull request closed without merging**